apps/prairielearn/src/ee/auth/saml/index.ts

```diff
@@ -1,77 +1,93 @@
-import { MultiSamlStrategy } from '@node-saml/passport-saml';
+import { MultiSamlStrategy, type SamlConfig } from '@node-saml/passport-saml';
 
 import { getInstitutionSamlProvider } from '../../lib/institution';
 
-export const strategy = new MultiSamlStrategy(
-  {
-    passReqToCallback: true,
-    getSamlOptions(req, done) {
-      getInstitutionSamlProvider(req.params.institution_id)
-        .then((samlProvider) => {
-          if (!samlProvider) {
-            return done(new Error('No SAML provider found for given institution'));
-          }
+export async function getSamlOptions({
+  institution_id,
+  host,
+  strictMode,
+}: {
+  institution_id: string;
+  host: string | undefined;
+  strictMode: boolean;
+}): Promise<SamlConfig> {
+  const samlProvider = await getInstitutionSamlProvider(institution_id);
+  if (!samlProvider) throw new Error('No SAML provider found for given institution');
+
+  // It's most convenient if folks can pass in `req.headers.host` directly,
+  // but that's typed as `string | undefined`. So, we'll accept that type
+  // and throw an error on on the off-change it's undefined.
+  if (!host) throw new Error('Missing host header');
```

Comment:
I'm fine with this as is, but we could also require callers to have processed `req.headers` through `zod` to get a type safe value.

Reply:
That feels like it's stretching the limits of what we should do with Zod tbh 😬 I'll keep this as is.